### PR TITLE
impl(generator): handle paths with multiple variables

### PR DESCRIPTION
**Pull request by scotthart**
part of the work for #12112 

The generator has been correctly handling url paths specified in the google.api.http option that only had one variable, e.g.:
```
    option (google.api.http) = {
      get: "/v1/{name=projects/*/serviceAccounts/*}"
    };
```
but not handling url paths that contain multiple variables, e.g.:
```
    option (google.api.http) = {
      post: "/v1/projects/{project}/instances/{instance}/addServerCa"
    };
```
(Currently, all google.api.http paths are formatted as one of these two examples.)

This PR adds support for the second case which affects both metadata contents for gRPC, as well as correctly identifying ALL path fields for REST that need not be added as query parameters.

The PR consists of one commit for the generator changes and a second commit for all the regenerated service files.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12162)

### CODE REVIEW

File: generator/internal/http_option_utils.cc [hunk 1]
```diff
@@ -242,21 +252,37 @@ ParseHttpExtension(google::protobuf::MethodDescriptor const& method) {
                      << ": google::api::HttpRule not handled";
   }
 
-  static std::regex const kUrlPatternRegex(R"((.*)\{(.*)=(.*)\}(.*))");
-  std::smatch match;
-  if (!std::regex_match(url_pattern, match, kUrlPatternRegex)) {
+  static std::regex const kImplicitNamedFieldUrlPatternRegex(
+      R"((.*)\{(.*)\}(.*))");
```
Comment (coryan) [inline generator/internal/http_option_utils.cc:256]: Can you have something like: `"/v1/{parent=projects/*/instances/*}/databases/{bar}/baz"`?

Is there a full grammar for these expressions we are trying to match?

File: generator/internal/http_option_utils_test.cc [hunk 2]
```diff
@@ -201,6 +208,23 @@ char const* const kServiceProto =
...
+    "       post: "
+    "\"/v1/projects/{project}/databases\"\n"
+    "       body: \"body\"\n"
+    "    };\n"
+    "  }\n"
+    "  // Leading comments about rpc Method7.\n"
+    "  rpc Method7(Implicit) returns (Empty) {\n"
+    "    option (google.api.http) = {\n"
+    "       post: "
+    "\"/v1/projects/{project}/instances/{instance}/databases\"\n"
```
Comment (coryan) [inline generator/internal/http_option_utils_test.cc:223]: Do we need a test using both the `{parent=...}` and `foo/{bar}/baz` syntax?

File: generator/internal/http_option_utils.cc [hunk 3]
```diff
@@ -242,21 +252,37 @@ ParseHttpExtension(google::protobuf::MethodDescriptor const& method) {
...
+  std::smatch implicit_match;
+
+  // The implicit named field regex matches both the implicit and explicit
```
Comment (coryan) [inline generator/internal/http_option_utils.cc:259]: Note to self: this comment was useful, don't miss it next time.

File: generator/internal/http_option_utils.cc [hunk 4]
```diff
@@ -242,21 +252,37 @@ ParseHttpExtension(google::protobuf::MethodDescriptor const& method) {
...
+    info.body = http_rule.body();
+    info.path_suffix = implicit_match[3];
+  }
 
   std::size_t current = 0;
-  for (auto open = url_pattern.find('{'); open != std::string::npos;
-       open = url_pattern.find('{', current)) {
+  auto open = url_pattern.find('{');
+  info.path_prefix = url_pattern.substr(current, open - current);
+  for (; open != std::string::npos; open = url_pattern.find('{', current)) {
```
Comment (coryan) [inline generator/internal/http_option_utils.cc:285]: I think there is a code smell here.  We are mixing regexp and explicit parsing.  That suggests we are doing something odd, like the grammar we want to parse is too complex for regex, but we don't want to write a full parser.

🤷

**Review by coryan — APPROVED** [review]
Assuming both implicit and explicit annotations can never appear in the same RPC, this LGTM. Otherwise maybe we should IM.

File: generator/internal/http_option_utils.cc (hunk 1 above)
Comment (coryan) [reply generator/internal/http_option_utils.cc:256]: I found it:

https://github.com/googleapis/googleapis/blob/02aeb715005c633115bc77d563d6920731c98189/google/api/http.proto#L227-L234

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12162)* status: 0 of 360 files reviewed, 4 unresolved discussions (waiting on @coryan)

*[a discussion](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12162#gh-1268908342:-N_t-sOi6o_0hSaJiCqa:bb5okle) (no related file):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I found it:

https://github.com/googleapis/googleapis/blob/02aeb715005c633115bc77d563d6920731c98189/google/api/http.proto#L227-L234

</blockquote></details>

Technically the grammar allows for the path you described. However, in practice that does not occur. I wrote a quick function to audit all the existing paths and none of them mix "Variable" formats (per their grammar). I also created an issue to add a check for unsupported grammars to this function and error if we encounter one https://github.com/googleapis/google-cloud-cpp/issues/12197

___
*[a discussion](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12162#gh-1268915962:-N_t0XNzEhIWPol9lLg_:b-vpxesn) (no related file):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map this GitHub inline comment thread to the right spot — sorry!)*

I think there is a code smell here.  We are mixing regexp and explicit parsing.  That suggests we are doing something odd, like the grammar we want to parse is too complex for regex, but we don't want to write a full parser.

🤷 
</blockquote></details>

Agreed. We could write a full parser, but there's no current usage of the grammar that exceeds our simple implementation of it. I will create an issue to consider adding a full parser in case it's something we want to tackle in the future https://github.com/googleapis/google-cloud-cpp/issues/12196

___

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12162)* status: 0 of 360 files reviewed, 4 unresolved discussions (waiting on @coryan)

*[a discussion](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12162#gh-1268910845:-N_t2JIo4hb1f0Wli8Uf:b-uyrlow) (no related file):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map this GitHub inline comment thread to the right spot — sorry!)*

Do we need a test using both the `{parent=...}` and `foo/{bar}/baz` syntax?
</blockquote></details>

It should be added as part of the solution for #12197 

___

**Review by coryan — COMMENTED** [review]
:lgtm:

> Reviewed 1 of 360 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12162#-:-N_t2ZXZ0jOj864CdgyQ:bnfp4nl)* status: 1 of 360 files reviewed, all discussions resolved

**Pull request merged**